### PR TITLE
Create retry buttons for Single task and full databricks jobs

### PR DESCRIPTION
**Pull request by dimberman**
# Description
This pull request adds functionality to the Airflow databricks workflow taskgroup to create "See Databricks Job Run", "Repair All Failed Tasks", and "Repair a single failed task" buttons. These buttons will allow users to easily navigate to the Databricks job run page and repair any failed tasks.

In the "job.launch" task, we add a "Repair All Failed Tasks" button because there is no way to create extra_links in a taskgroup. This button will launch a repair request to Databricks via the JobService API endpoint to repair all failed tasks in the job.

Screenshot 2023-02-26 at 10 22 48 PM

When the repair buttons are clicked, a repair request is sent to Databricks via the JobService API endpoint. After that we flash a message in the UI to let the user know that the repair has started along with the repair id.

Screenshot 2023-02-26 at 10 30 12 PM

In the notebook tasks, we add a "Repair a single failed task" button that allows users to rerun just the associated task. This button will also launch a repair request to Databricks via the JobService API endpoint.

# Reasoning

We originally attempted to use Airflow's built-in retry logic to repair failed tasks, but on trying this we ran into a core difference between how airflow and databricks handle retries. Databricks only allows one repair at a time, meaning that for each task airflow attempts to retry, databricks would start a new job with a new jobs cluster. To efficiently use databricks we need to send all repair requests in a single request, so we created the "Repair All Failed Tasks" and "Repair a single failed task" buttons.

### CODE REVIEW

File: cosmos/providers/databricks/plugin.py
Comment (tatiana) [inline cosmos/providers/databricks/plugin.py:48]: Should we be using standard libraries `logging` features?

File: docs/databricks/workflows.rst
Comment (tatiana) [inline docs/databricks/workflows.rst:36]: We need to update now that we have the repair of all failed tasks.
Do we have a GitHub ticket for the "repair failed & downstream tasks"?

**Comment by tatiana** [comment]
> (...)  because there is no way to create extra_links in a taskgroup

@dimberman, would it make sense for us to create an issue in Apache Airflow Github to add support to having `extra_links` in task groups as well? Or is there any blockers to be able to accomplish this? I understand this would only be available in future versions of Airflow, but it probably makes sense.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:48]: This is somethign I wanted to discuss. These classes don't inherit from LoggingMixin so wanted to figure out what best practices are for logging here.

**Comment by bbovenzi** [comment]
> 

I think that could be useful. Although I'm not too familiar with how that would work. We don't define much on a Task Groups right now, since its supposed to be mainly a UI concept. Aren't extra links connected to an operator?

**Comment by dimberman** [comment]
@bbovenzi for the moment they're associated with operators, but I don't see why we can't add them to taskgroups, especially extra_links since they're just UI buttons

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:125]:
```suggestion
```
2 returns :)

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:122]: Can we not use the below? It looks more convenient method as per Airflow's docstring for `get_one` implementation.
```suggestion
        metadata = XCom.get_value(
            ti_key=ti_key,
            key="return_value",
        )
```

Ref: [databricks/support-for-retries ](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/xcom/index.html#airflow.models.xcom.BaseXCom.get_value)

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:146]:
```suggestion
        metadata = XCom.get_one(
            ti_key=ti_key,
            key="return_value",
        )
```
same suggestion as above

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:182]: Do we need `up_for_retry` here? Do we see any possible race conditions here? e.g. A task that was up for retry gets retried by the Airflow scheduler just before a repair is triggered from here.

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:50]: Will all the needed tasks have this same task ID `f"{group_id}.launch"`?

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:267]: We will need suitable name and category values here.

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:273]: Is this the default and needed value? Or do we need to change it?

File: cosmos/providers/databricks/notebook.py
Comment (dimberman) [reply cosmos/providers/databricks/notebook.py:22]: Yes I can remove this thanks for catching that.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:122]: Oh neat didn't know this was a thing. Thanks @pankajkoti !

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:46]: Any databricks job will have a launch task, we don't want to relaunch so we don't add it to the repair.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:282]: We can change it I believe.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py]: Yes we need those too because we need to clear those tasks for them to actually be retried. If anything we might want to advise people to turn off retries for databricks taskgroup jobs or default the tasks to not retry. @tatiana @kaxil WDYT?

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:46]: The idea is _not_ to add the group_id.launch task. That's not a databricks task so we shouldn't submit it.

File: cosmos/providers/databricks/notebook.py
Comment (pankajkoti) [reply cosmos/providers/databricks/notebook.py]: I still feel that it needs to be >=1. 

Case 1: For the first record, len(repair_history) will be zero and this case is satisfied with both len(repair_history) >=1 and len(repair_history)>1
Case 2: For the second record, len(repair_history) will be 1 and this case won't be satisfied with len(repair_history) > 1
Case 3: For the third record, len(repair_history) will be 2 and this case will be satisfied with len(repair_history) > 1

File: cosmos/providers/databricks/plugin.py
Comment (pankajkoti) [inline cosmos/providers/databricks/plugin.py:48]:
```suggestion
        if task_id == f"{group_id}.launch":
            continue
        print(f"databricks task id {_get_databricks_task_id(task)}")
        task_ids.append(_get_databricks_task_id(task))
```

File: cosmos/providers/databricks/notebook.py
Comment (pankajkoti) [reply cosmos/providers/databricks/notebook.py]: I checked the API. len(repair_history) is never zero, and the repair history starts with the original record. So I believe what you're saying is right.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:46]: @pankajkoti we don't need to worry about that as we are only looking at tasks within that task group.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py:46]: @tatiana I was able to set up logging using the LoggingMixin class for Airflow

Comment (dimberman) [reply cosmos/providers/databricks/plugin.py]: fixed this by using LoggingMixin

**Review by pankajkoti — APPROVED** [review]
Mostly looks good to me. One question I have is, once a repair request is sent to Databricks for a failed task(s), do we monitor the status of the task(s)? I mean if a failed task is repaired successfully will it's status change to success in the Airflow UI? 
If we're not handling that at the moment, can we log a ticket for that to check the feasibility for that?

**Review by tatiana — APPROVED** [review]
I really liked the plugin strategy, @dimberman , and UI change. One last minor comment on a open thread.

We currently support "apache-airflow>=2.4" and Python 3.7-3.10. Should we log a ticket to ensure we're running the tests/example DAGs against these combinations? Perhaps we can add this to #161, WDYT? cc @pankajkoti 

I'm slightly worried we are introducing 747 lines and that we are not running the tests in the CI yet, but relieved that @pankajkoti is working on it!

File: cosmos/providers/databricks/plugin.py
Comment (tatiana) [reply cosmos/providers/databricks/plugin.py]: Do we really need `up_for_retry` here, @dimberman ?

**Comment by dimberman** [comment]
@pankajkoti by clearing the tasks, the airflow scheduler will pick them up as if they've never run and monitor them. 

@tatiana i agree we need CI/CD ASAP. Once @pankajkoti merges his PR, we can address any issues that come up. I don't think anything I've added here is specific to new versions of airflow or python.

**Comment by dimberman** [comment]
Ok as a final test I just ran this against a local airflow using `astro dev start` and can confirm that running the databricks DAG as well as the repairs and job links all work correctly.

File: cosmos/providers/databricks/plugin.py
Comment (dimberman) [reply cosmos/providers/databricks/plugin.py]: @tatiana yes, because we can't use Airflow's retries so any task that is "up_for_retry" will just continue to fail unless it is added to the repair job.

**Pull request merged**
